trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/YamlCorrectnessWitnessExtractor.java

```diff
@@ -0,0 +1,121 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator.witness;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.eclipse.cdt.core.dom.ast.ASTGenericVisitor;
+import org.eclipse.cdt.core.dom.ast.IASTFileLocation;
+import org.eclipse.cdt.core.dom.ast.IASTNode;
+import org.eclipse.cdt.core.dom.ast.IASTTranslationUnit;
+
+import de.uni_freiburg.informatik.ultimate.core.model.services.IUltimateServiceProvider;
+import de.uni_freiburg.informatik.ultimate.util.datastructures.relation.HashRelation;
+import de.uni_freiburg.informatik.ultimate.witnessparser.yaml.LocationInvariant;
+import de.uni_freiburg.informatik.ultimate.witnessparser.yaml.LoopInvariant;
+import de.uni_freiburg.informatik.ultimate.witnessparser.yaml.Witness;
+import de.uni_freiburg.informatik.ultimate.witnessparser.yaml.WitnessEntry;
+
+public class YamlCorrectnessWitnessExtractor extends CorrectnessWitnessExtractor {
+	private Witness mWitness;
+
+	public YamlCorrectnessWitnessExtractor(final IUltimateServiceProvider service) {
+		super(service);
+	}
+
+	public void setWitness(final Witness witness) {
+		mWitness = witness;
+	}
+
+	@Override
+	public boolean isReady() {
+		return mWitness != null && mTranslationUnit != null;
+	}
+
+	@Override
+	protected HashRelation<IASTNode, IExtractedWitnessEntry> extract() {
+		final HashRelation<IASTNode, IExtractedWitnessEntry> rtr = new HashRelation<>();
+		for (final WitnessEntry entry : mWitness.getEntries()) {
+			int line = -1;
+			if (entry instanceof LocationInvariant && !mCheckOnlyLoopInvariants) {
+				line = ((LocationInvariant) entry).getLocation().getLine();
+			}
+			if (entry instanceof LoopInvariant) {
+				line = ((LoopInvariant) entry).getLocation().getLine();
+			}
```

Comment:
Is it possible that an  entry is a LocationInvariant and a LoopInvariant? 
What happens if the LineMatchingVisitor gets -1 as line?

Reply:
> Is it possible that an entry is a LocationInvariant and a LoopInvariant?

No

>  What happens if the LineMatchingVisitor gets -1 as line?

It skips the witness entry (so the result does not contain it), because there are (hopefully) no IASTNodes in that line.

Reply:
So maybe you could use if-else and else instead of the two checks?